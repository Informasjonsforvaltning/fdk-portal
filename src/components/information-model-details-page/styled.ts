import styled from 'styled-components';
import LinkBase from '@fellesdatakatalog/link';
import TabsBase from '@fellesdatakatalog/tabs';
import { theme, Colour } from '@fellesdatakatalog/theme';

const Link = styled(LinkBase)`
  color: ${theme.colour(Colour.VIOLET, 'V50')} !important;
`;

const Tabs = styled(TabsBase)`
  padding: ${theme.spacing('S16')};
  border-radius: 5px;
  background: ${theme.colour(Colour.NEUTRAL, 'N0')};
`;

const Tab = styled.span`
  padding: ${theme.spacing('S6')} ${theme.spacing('S12')};

  &.active,
  &:before,
  &:after {
    border-color: ${theme.colour(Colour.VIOLET, 'V30')};
  }

  &:not(.active) {
    color: ${theme.colour(Colour.VIOLET, 'V50')};
  }
`;

const Pane = styled.div`
  margin-top: ${theme.spacing('S12')};
`;

const Code = styled.pre`
  white-space: pre-wrap;
  overflow-wrap: break-word;
`;

export default { Link, Tabs, Tab, Pane, Code };
