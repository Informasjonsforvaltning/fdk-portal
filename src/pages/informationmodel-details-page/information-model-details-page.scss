.fdk-icon-catalog-infomodel {
  background-image: url("/img/icon-catalog-infomodel-lg.svg");
  width: 3.5em; 
  height: 3.5em;
  display: inline-block;
}
