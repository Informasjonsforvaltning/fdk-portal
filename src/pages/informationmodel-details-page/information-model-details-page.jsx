import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import DocumentMeta from 'react-document-meta';

import localization from '../../lib/localization';
import { getTranslateText } from '../../lib/translateText';
import { HarvestDate } from '../../components/harvest-date/harvest-date.component';
import { SearchHitHeader } from '../../components/search-hit-header/search-hit-header.component';
import { StickyMenu } from '../../components/sticky-menu/sticky-menu.component';
import { ListRegular } from '../../components/list-regular/list-regular.component';
import { Tabs } from '../../components/tabs/tabs.component';
import { Structure } from './structure/structure.component';
import './information-model-details-page.scss';

const renderJSONSchema = schema => {
  if (!schema) {
    return null;
  }

  return (
    <div>
      <pre style={{ whiteSpace: 'pre' }}>{JSON.stringify(schema, null, 2)}</pre>
    </div>
  );
};

const createTabsArray = schema => {
  const tabsArray = [];

  // only show structure-tab if schema is an object, not when schema is an array
  if (schema && schema.definitions) {
    tabsArray.push({
      title: localization.infoMod.tabs.structure,
      body: <Structure definitions={schema.definitions} />
    });
  }
  tabsArray.push({
    title: localization.infoMod.tabs.json,
    body: renderJSONSchema(schema)
  });
  return tabsArray;
};

const renderModels = schema => {
  if (!schema) {
    return null;
  }

  return (
    <ListRegular title={localization.infoMod.infoModHeader}>
      <div className="d-flex list-regular--item" />
      <Tabs tabContent={createTabsArray(schema)} />
    </ListRegular>
  );
};

const renderRelatedApi = (referencedApis, publisherItems) => {
  if (!referencedApis || referencedApis.length === 0) {
    return null;
  }
  const apiReferenceItems = items =>
    items.map((item, index) => (
      <div key={`reference-${index}`} className="list-regular--item">
        <SearchHitHeader
          title={item.title}
          titleLink={`/apis/${encodeURIComponent(item.id)}`}
          publisherLabel={`${localization.responsible}:`}
          publisher={item.publisher}
          publisherItems={publisherItems}
          tag="h4"
          darkThemeBackground
        />
      </div>
    ));

  return (
    <ListRegular title={localization.infoMod.relatedAPIHeader}>
      {apiReferenceItems(referencedApis)}
    </ListRegular>
  );
};

const renderStickyMenu = (informationModelItem, referencedApis) => {
  const menuItems = [];

  if (_.get(informationModelItem, 'schema')) {
    menuItems.push({
      name: localization.infoMod.infoModHeader,
      prefLabel: localization.infoMod.infoModHeader
    });
  }
  if (!_.isEmpty(referencedApis)) {
    menuItems.push({
      name: localization.infoMod.relatedAPIHeader,
      prefLabel: localization.infoMod.relatedAPIHeader
    });
  }

  return <StickyMenu title={localization.goTo} menuItems={menuItems} />;
};

function getSchema(model) {
  const schemaJson = _.get(model, 'schema');
  return JSON.parse(schemaJson);
}

export const InformationModelDetailsPage = ({
  fetchPublishersIfNeeded,
  informationModelItem,
  publisherItems,
  referencedApis
}) => {
  fetchPublishersIfNeeded();

  if (!informationModelItem) {
    return null;
  }

  const meta = {
    title: getTranslateText(informationModelItem.title),
    description: getTranslateText(informationModelItem.description)
  };

  return (
    <main id="content" className="container">
      <article>
        <div className="row">
          <div className="col-12 col-lg-9 offset-lg-3">
            <DocumentMeta {...meta} />

            <div className="d-flex align-items-center fdk-detail-date mb-5">
              <i className="align-self-center fdk-icon-catalog-infomodel mr-2" />
              <strong className="align-self-center">
                {localization.infoMod.informationModelDescription}&nbsp;
              </strong>
              <HarvestDate
                className="align-self-center"
                harvest={informationModelItem.harvest}
              />
            </div>
          </div>
        </div>

        <div className="row">
          <div className="col-12 col-lg-3 ">
            {renderStickyMenu(informationModelItem, referencedApis)}
          </div>

          <section className="col-12 col-lg-9 mt-3">
            <SearchHitHeader
              title={getTranslateText(informationModelItem.title)}
              publisherLabel={`${localization.responsible}:`}
              publisher={informationModelItem.publisher}
              publisherItems={publisherItems}
              nationalComponent={informationModelItem.nationalComponent}
              darkThemeBackground
            />

            {renderModels(getSchema(informationModelItem))}

            {renderRelatedApi(referencedApis, publisherItems)}
            <div style={{ height: '75vh' }} />
          </section>
        </div>
      </article>
    </main>
  );
};

InformationModelDetailsPage.defaultProps = {
  informationModelItem: null,
  publisherItems: null,
  fetchPublishersIfNeeded: () => {},
  referencedApis: []
};

InformationModelDetailsPage.propTypes = {
  informationModelItem: PropTypes.object,
  publisherItems: PropTypes.object,
  fetchPublishersIfNeeded: PropTypes.func,
  referencedApis: PropTypes.array
};
