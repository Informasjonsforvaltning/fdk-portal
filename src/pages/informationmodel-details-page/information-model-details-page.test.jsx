import React from 'react';
import { shallow } from 'enzyme';
import { InformationModelDetailsPage } from './information-model-details-page';
import informationModelApiResponse from '../../mock/informationmodel.response.json';

let wrapper;

beforeEach(() => {
  const fetchPublishersIfNeeded = jest.fn();

  const defaultProps = {
    showFilterModal: true,
    informationModelItem: informationModelApiResponse,
    fetchPublishersIfNeeded
  };
  wrapper = shallow(<InformationModelDetailsPage {...defaultProps} />);
});

test('should render InformationModelDetailsPage correctly with no informationModelItem', () => {
  const wrapperWithNoProps = shallow(<InformationModelDetailsPage />);
  expect(wrapperWithNoProps).toMatchSnapshot();
});

test('should render InformationModelDetailsPage correctly with informationModelItem', () => {
  expect(wrapper).toMatchSnapshot();
});
