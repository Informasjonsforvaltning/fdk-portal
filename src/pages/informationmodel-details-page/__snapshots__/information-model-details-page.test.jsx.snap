// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render InformationModelDetailsPage correctly with informationModelItem 1`] = `
<main
  className="container"
  id="content"
>
  <article>
    <div
      className="row"
    >
      <div
        className="col-12 col-lg-9 offset-lg-3"
      >
        <SideEffect(DocumentMeta)
          description={null}
          title="Åpne Data fra Enhetsregisteret - API Dokumentasjon"
        />
        <div
          className="d-flex align-items-center fdk-detail-date mb-5"
        >
          <i
            className="align-self-center fdk-icon-catalog-infomodel mr-2"
          />
          <strong
            className="align-self-center"
          >
            Information Model Description
             
          </strong>
          <HarvestDate
            className="align-self-center"
            harvest={
              Object {
                "changed": Array [
                  "2019-01-28T11:31:59.499+0000",
                  "2019-01-28T13:07:29.496+0000",
                ],
                "firstHarvested": "2019-01-26T12:11:59.193+0000",
                "lastChanged": "2019-01-28T13:07:29.496+0000",
                "lastHarvested": "2019-01-28T13:07:29.496+0000",
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12 col-lg-3 "
      >
        <StickyMenu
          menuItems={
            Array [
              Object {
                "name": "Information model",
                "prefLabel": "Information model",
              },
            ]
          }
          title="Go to"
        />
      </div>
      <section
        className="col-12 col-lg-9 mt-3"
      >
        <SearchHitHeader
          darkThemeBackground={true}
          externalLink={false}
          nationalComponent={false}
          publisher={null}
          publisherItems={null}
          publisherLabel="Responsible:"
          publisherTag="span"
          referenceData={null}
          statusCode={null}
          tag="h1"
          theme={Array []}
          title="Åpne Data fra Enhetsregisteret - API Dokumentasjon"
          titleLink={null}
        />
        <ListRegular
          bottomMargin={true}
          title="Information model"
        >
          <div
            className="d-flex list-regular--item"
          />
          <Tabs
            tabContent={
              Array [
                Object {
                  "body": <Structure
                    definitions={
                      Object {
                        "Adresse": Object {
                          "properties": Object {
                            "adresse": Object {
                              "items": Object {
                                "type": "string",
                              },
                              "type": "array",
                            },
                            "kommune": Object {
                              "type": "string",
                            },
                            "kommunenummer": Object {
                              "type": "string",
                            },
                            "land": Object {
                              "type": "string",
                            },
                            "landkode": Object {
                              "type": "string",
                            },
                            "postnummer": Object {
                              "type": "string",
                            },
                            "poststed": Object {
                              "type": "string",
                            },
                          },
                        },
                        "Enhet": Object {
                          "properties": Object {
                            "_links": Object {
                              "$ref": "#/definitions/Self",
                            },
                            "antallAnsatte": Object {
                              "type": "integer",
                            },
                            "forretningsadresse": Object {
                              "$ref": "#/definitions/Adresse",
                            },
                            "institusjonellSektorkode": Object {
                              "$ref": "#/definitions/Sektorkode",
                            },
                            "konkurs": Object {
                              "type": "boolean",
                            },
                            "maalform": Object {
                              "type": "string",
                            },
                            "naeringskode1": Object {
                              "$ref": "#/definitions/Naeringskode",
                            },
                            "naeringskode2": Object {
                              "$ref": "#/definitions/Naeringskode",
                            },
                            "naeringskode3": Object {
                              "$ref": "#/definitions/Naeringskode",
                            },
                            "navn": Object {
                              "type": "string",
                            },
                            "organisasjonsform": Object {
                              "$ref": "#/definitions/Organisasjonsform",
                            },
                            "organisasjonsnummer": Object {
                              "type": "string",
                            },
                            "postadresse": Object {
                              "$ref": "#/definitions/Adresse",
                            },
                            "registreringsdatoEnhetsregisteret": Object {
                              "type": "string",
                            },
                            "registrertIForetaksregisteret": Object {
                              "type": "boolean",
                            },
                            "registrertIFrivillighetsregisteret": Object {
                              "type": "boolean",
                            },
                            "registrertIMvaregisteret": Object {
                              "type": "boolean",
                            },
                            "registrertIStiftelsesregisteret": Object {
                              "type": "boolean",
                            },
                            "sisteInnsendteAarsregnskap": Object {
                              "type": "string",
                            },
                            "stiftelsedato": Object {
                              "type": "string",
                            },
                            "underAvvikling": Object {
                              "type": "boolean",
                            },
                            "underTvangsavviklingEllerTvangsopplosning": Object {
                              "type": "boolean",
                            },
                          },
                        },
                        "Href": Object {
                          "properties": Object {
                            "href": Object {
                              "type": "string",
                            },
                          },
                        },
                        "Naeringskode": Object {
                          "properties": Object {
                            "beskrivelse": Object {
                              "type": "string",
                            },
                            "kode": Object {
                              "type": "string",
                            },
                          },
                        },
                        "Organisasjonsform": Object {
                          "properties": Object {
                            "_links": Object {
                              "$ref": "#/definitions/Self",
                            },
                            "beskrivelse": Object {
                              "type": "string",
                            },
                            "kode": Object {
                              "type": "string",
                            },
                          },
                        },
                        "Organisasjonsformer": Object {
                          "properties": Object {
                            "organisasjonsformer": Object {
                              "items": Object {
                                "$ref": "#/definitions/Organisasjonsform",
                              },
                              "type": "array",
                            },
                          },
                        },
                        "Sektorkode": Object {
                          "properties": Object {
                            "beskrivelse": Object {
                              "type": "string",
                            },
                            "kode": Object {
                              "type": "string",
                            },
                          },
                        },
                        "Self": Object {
                          "properties": Object {
                            "self": Object {
                              "$ref": "#/definitions/Href",
                            },
                          },
                        },
                        "_Organisasjonsformer": Object {
                          "properties": Object {
                            "embedded": Object {
                              "$ref": "#/definitions/Organisasjonsformer",
                            },
                          },
                        },
                      }
                    }
                  />,
                  "title": "Structure",
                },
                Object {
                  "body": <div>
                    <pre
                      style={
                        Object {
                          "whiteSpace": "pre",
                        }
                      }
                    >
                      {
  "$schema": "http://json-schema.org/draft-06/schema#",
  "$id": "https://fellesdatakatalog.brreg.no/informationmodels/936c581f-9d95-418b-8fce-e452f6bdfd5e/schema",
  "definitions": {
    "Enhet": {
      "properties": {
        "organisasjonsnummer": {
          "type": "string"
        },
        "navn": {
          "type": "string"
        },
        "organisasjonsform": {
          "$ref": "#/definitions/Organisasjonsform"
        },
        "postadresse": {
          "$ref": "#/definitions/Adresse"
        },
        "registreringsdatoEnhetsregisteret": {
          "type": "string"
        },
        "registrertIMvaregisteret": {
          "type": "boolean"
        },
        "naeringskode1": {
          "$ref": "#/definitions/Naeringskode"
        },
        "naeringskode2": {
          "$ref": "#/definitions/Naeringskode"
        },
        "naeringskode3": {
          "$ref": "#/definitions/Naeringskode"
        },
        "antallAnsatte": {
          "type": "integer"
        },
        "forretningsadresse": {
          "$ref": "#/definitions/Adresse"
        },
        "stiftelsedato": {
          "type": "string"
        },
        "institusjonellSektorkode": {
          "$ref": "#/definitions/Sektorkode"
        },
        "registrertIForetaksregisteret": {
          "type": "boolean"
        },
        "registrertIStiftelsesregisteret": {
          "type": "boolean"
        },
        "registrertIFrivillighetsregisteret": {
          "type": "boolean"
        },
        "sisteInnsendteAarsregnskap": {
          "type": "string"
        },
        "konkurs": {
          "type": "boolean"
        },
        "underAvvikling": {
          "type": "boolean"
        },
        "underTvangsavviklingEllerTvangsopplosning": {
          "type": "boolean"
        },
        "maalform": {
          "type": "string"
        },
        "_links": {
          "$ref": "#/definitions/Self"
        }
      }
    },
    "Adresse": {
      "properties": {
        "land": {
          "type": "string"
        },
        "landkode": {
          "type": "string"
        },
        "postnummer": {
          "type": "string"
        },
        "poststed": {
          "type": "string"
        },
        "adresse": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "kommune": {
          "type": "string"
        },
        "kommunenummer": {
          "type": "string"
        }
      }
    },
    "Naeringskode": {
      "properties": {
        "kode": {
          "type": "string"
        },
        "beskrivelse": {
          "type": "string"
        }
      }
    },
    "Sektorkode": {
      "properties": {
        "kode": {
          "type": "string"
        },
        "beskrivelse": {
          "type": "string"
        }
      }
    },
    "_Organisasjonsformer": {
      "properties": {
        "embedded": {
          "$ref": "#/definitions/Organisasjonsformer"
        }
      }
    },
    "Organisasjonsformer": {
      "properties": {
        "organisasjonsformer": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/Organisasjonsform"
          }
        }
      }
    },
    "Organisasjonsform": {
      "properties": {
        "kode": {
          "type": "string"
        },
        "beskrivelse": {
          "type": "string"
        },
        "_links": {
          "$ref": "#/definitions/Self"
        }
      }
    },
    "Self": {
      "properties": {
        "self": {
          "$ref": "#/definitions/Href"
        }
      }
    },
    "Href": {
      "properties": {
        "href": {
          "type": "string"
        }
      }
    }
  }
}
                    </pre>
                  </div>,
                  "title": "JSON Schema",
                },
              ]
            }
          />
        </ListRegular>
        <div
          style={
            Object {
              "height": "75vh",
            }
          }
        />
      </section>
    </div>
  </article>
</main>
`;

exports[`should render InformationModelDetailsPage correctly with no informationModelItem 1`] = `""`;
